### ISSUE TITLE
Parameter in script is replaced with environment variable instead of configured parameter

### ISSUE THREAD

**Issue by sebtacke** [liquibase/liquibase#1644]
## Environment
**Liquibase Version**: 3.8.9
(could reproduce issue with 3.10.3 and 4.2.2)

**Liquibase Integration & Version**: spring boot

**Liquibase Extension(s) & Version**: N/A

**Database Vendor & Version**: Microsoft SQL Server 2017, Postgres 11.6
(issue is not db specific)

**Operating System Type & Version**: Linux 5.10.9

## Description
${user} was replaced with the value of the USER environment variable instead of the value configured in spring.liquibase.parameters.user

## Steps To Reproduce
* Write custom SQL migration script with parameter placeholder
* Include script in liquibase changelog for a spring boot application
* Start spring boot application
  * Set value for parameter via spring.liquibase.parameters...
  * Set environment variable with parameter name and different value

## Actual Behavior
The parameter placeholder is replaced with the value of the environment variable.

## Expected/Desired Behavior
The parameter placeholder is replaced with the explicitly configured value in spring.liquibase.parameters...

## Screenshots (if appropriate)
N/A

## Additional Context
The spring integration constructs a liquibase instance which reads all system properties and environment variables into the change log parameters in the constructor. Afterwards it sets the configured parameters via the setChangeLogParameter method.
This method doesn't overwrite global parameters that are already present, so any parameter from system properties or environment variables will block the configured parameters.

**Comment by sebtacke** [comment]
In my oppinion this is a larger issue than just the priority of the parameter sources. The system properties and environment variables are merged with the explicitly configured parameters without any isolation.
There's no source information for the parameters and the properties/variables are not limited to ones with a specific prefix/namespace. So you have little control over what's included in the change log parameters.
It's also not possible to control/configure whether you actually want to include system properties and/or environment variables in addition to the explicitly configured parameters.

**Comment by molivasdat** [comment]
Hi @sebtacke 
Thanks for writing up this issue.

Here is the order of precendence as specified in  SpringBoot documentation
https://docs.spring.io/spring-boot/docs/1.2.3.RELEASE/reference/html/boot-features-external-config.html
    

1.      Command line arguments.
2.     JNDI attributes from java:comp/env.
3.     Java System properties (System.getProperties()).
4.     OS environment variables.
5.     A RandomValuePropertySource that only has properties in random.*.
6.     Profile-specific application properties outside of your packaged jar (application-{profile}.properties and YAML variants)
7.     Profile-specific application properties packaged inside your jar (application-{profile}.properties and YAML variants)
8.     Application properties outside of your packaged jar (application.properties and YAML variants).
9.     Application properties packaged inside your jar (application.properties and YAML variants).
10.     @PropertySource annotations on your @Configuration classes.
11.     Default properties (specified using SpringApplication.setDefaultProperties).

**Comment by sebtacke** [comment]
I know the spring boot configuration order. It would apply if the environment variable was SPRING_LIQUIBASE_PARAMETERS_USER. However what happened here is not related to spring boot at all. The environment variable that overrode the explicitly configured parameter was the USER environment variable that is set by the Linux shell to the name of the currently loged in user.

As far as I can tell the USER environment variable would also override the SPRING_LIQUIBASE_PARAMETERS_USER or -Dspring.liquibase.parameters.user system property.

The ChangeLogParameters class loads initial parameters from the environment and the system properties:
https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/changelog/ChangeLogParameters.java#L31-L40
These are put into the change log parameter list as global parameters directly. This is called from the Liquibase constructor, so it happens before the spring configuration is applied.

Then the SpringLiquibase integration tries to set the explicitly configured parameters:
https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/integration/spring/SpringLiquibase.java#L331-L335

However the ChangeLogParameters will not modify parameters that are already present in the list:
https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/changelog/ChangeLogParameters.java#L100-L110

Any environment variable that is present can overwrite your explicit configuration. That should never be allowed to happen.

**Comment by nvoxland** [comment]
Yes, that does need to be improved. What we are trying to support is users being able to set properties via different mechanisms, like a CLI user can add an environment variable and use that for a changelog property.

That capability grew organically over the years, and we are actually currently in the process of tightening all that up, planning for it coming out with Liquibase 4.4.

In the meantime, you are hitting the problem particularly because your user setting is conflicting with the standard USER environment variable. Are you able to change to something like `spring.liquibase.parameters.dbuser` and `${dbuser}` instead?

**Comment by sebtacke** [comment]
I already tried this work-around and it worked for a fresh install. It would be really ugly for existing installations though, because it changes the script's checksum.

The issue occurred due to a Spring Boot update that updated the liquibase version from 3.6.x to 3.8.x. The older version didn't read the envrioment. I also tried downgrading the liquibase version and using it with the newer Spring Boot version, but that didn't work, because it needs the clear checksums config.

We've postponed the Spring Boot update for now. Depending on the timeline for the fix I might patch the ChangeLogParameters class. I think I'd rather remove a small part of the functionality we don't use, than mess with the checksums for already executed scripts in the production environments.
We also have additional liquibase parameters and they might be compromised on other systems as we don't have full control over the process's environment.

**Comment by sebtacke** [comment]
Regarding setting properties via different mechanisms from CLI or other frameworks:
I think the core lib should never do that. It's very prone to break downstream projects. The mechanism should be opt-in only. For instance providing two methods in the Liquibase class: One that reads the environment into the parameters and another that reads the system properties into the parameters.
That way you would have to call two extra methods in the CLI implementation, but all other clients can now control whether they want to include them and in which order.
Including an optional prefix parameter for loading the properties and environment variables might be even better, because it would allow downstream projects to use the mechanism and fit the variables/properties into their own naming scheme. On the other hand this might be unnecessary as projects like Spring Boot already have full fledged config support and you only need the setChangeLogParameter method for them.

**Comment by FBurguer** [comment]
This is not an issue anymore as explained by Nathan on PR #[2157](https://github.com/liquibase/liquibase/pull/2157), he made a change on the docs page to explain how to work around it [here](https://docs.liquibase.com/concepts/changelogs/property-substitution.html) look for `${: property-name}`. Thanks!

**Comment by sebtacke** [comment]
If I understand the documentation correctly this doesn't actually fix my issue @FBurguer. I don't want to write `${user}` into the resulting SQL script. I want to replace `${user}` with the configured value from the Spring Boot configuration for the parameters https://github.com/liquibase/liquibase/blob/master/liquibase-core/src/main/java/liquibase/integration/spring/SpringLiquibase.java#L66 but it's replaced with the environment variable user instead.
https://github.com/liquibase/liquibase/pull/2157 would have fixed the issue.

**Comment by AQS-DTheuke** [comment]
@FBurguer Which version adds the parameter substitution escaping?
I'm using 4.3.5 and if I use `${: user}` in the changelogs, then `${: user}` will be in my database.

**Comment by FBurguer** [comment]
@sebtacke Thanks for the response! @nvoxland belives that #2157 does not fix this issue but it can help finding the solution to it. So we are adding this issue to the list of open issues to fix  and get reviewed and prioritzed. You are always welcome to submit a PR! Thanks.

**Comment by FBurguer** [comment]
@AQS-DTheuke Hi! Did you try upgrading to our latest version?

**Comment by AQS-DTheuke** [comment]
Yes, we also tried with v4.17.0.
